plugins/weida-raw/fu-weida-raw-device.c

```diff
@@ -0,0 +1,1093 @@
+#define FU_WEIDA_RAW_PAGE_SIZE		  0x1000
+#define FU_WEIDA_RAW_FLASH_PAGE_SIZE	  256
+#define FU_WEIDA_RAW_USB_MAX_PAYLOAD_SIZE 63
+
+#define FU_WEIDA_RAW_FOURCC_ID_RIFF 0x46464952
+#define FU_WEIDA_RAW_FOURCC_ID_WIF2 0x32464957
+#define FU_WEIDA_RAW_FOURCC_ID_WHIF 0x46494857
+#define FU_WEIDA_RAW_FOURCC_ID_INFO 0x4f464e49
+#define FU_WEIDA_RAW_FOURCC_ID_FSUM 0x4d555346
+#define FU_WEIDA_RAW_FOURCC_ID_FERA 0x41524546
+#define FU_WEIDA_RAW_FOURCC_ID_FBIN 0x4e494246
+#define FU_WEIDA_RAW_FOURCC_ID_FRMT 0x544D5246
+#define FU_WEIDA_RAW_FOURCC_ID_FRWR 0x52575246
+#define FU_WEIDA_RAW_FOURCC_ID_CNFG 0x47464E43
+
+typedef struct {
+	guint32 FourCC;
+	guint32 FileSize;
+	guint32 DataType;
+} FuWeidaRiffHeader;
+
+typedef struct {
+	guint32 FourCC;
+	guint32 Size; // payload size
+} FuWeidaChunkHeader;
+
+typedef struct {
+	guint32 Address;
+	guint32 Size;
+} FuWeidaFlashSpace;
+
+typedef struct {
+	guint32 Address;
+	guint32 Size;
+	guint8 *bin;
+} FuWeidaSpiBinary;
+
+typedef enum {
+	FU_WDT8760_COMMAND9 = 0x06,
+	FU_WDT8760_COMMAND63 = 0x07,
+	FU_WDT8760_MODE_FLASH_PROGRAM = 0x96,
+	FU_WDT8760_READ_BUFFERED_RESPONSE = 0xC7,
+	FU_WDT8760_GET_DEVICE_STATUS = 0xC9,
+	FU_WDT8760_SET_DEVICE_MODE = 0xCA,
+	FU_WDT8760_REBOOT = 0xCE,
+	FU_WDT8760_SET_FLASH_ADDRESS = 0xD0,
+	FU_WDT8760_READ_FLASH = 0xD1,
+	FU_WDT8760_ERASE_FLASH = 0xD2,
+	FU_WDT8760_WRITE_FLASH = 0xD3,
+	FU_WDT8760_PROTECT_FLASH = 0xD4,
+	FU_WDT8760_CALCULATE_FLASH_CHECKSUM = 0xD5,
+
+} FU_WDT8760Command;
+
+typedef enum {
+	FU_WDT8760_UnprotectLower508k = 0x0044,
+	FU_WDT8760_ProtectAll512k = 0x007C,
+} FU_WDT8760_FlashProtect;
+
+#define FU_WEIDA_RAW_FW_NOT_SUPPORT 0x00
+#define FU_WEIDA_RAW_FW_MAYBE_ISP   0x01
+#define FU_WEIDA_RAW_FW8755	    0x02
+#define FU_WEIDA_RAW_FW8760	    0x4
+#define FU_WEIDA_RAW_FW8762	    0x08
+#define FU_WEIDA_RAW_FW8790	    0X10
+
+struct _FuWeidaRawDevice {
+	FuUdevDevice parent_instance;
+	gint32 dev_type;
+	gint32 firmware_id;
+	gint32 hardware_id;
+	gint32 serial_number;
+	guint8 firmware_rev_ext;
+};
+
+G_DEFINE_TYPE(FuWeidaRawDevice, fu_weida_raw_device, FU_TYPE_UDEV_DEVICE)
+
+static gboolean
+fu_weida_raw_device_set_feature(FuWeidaRawDevice *self,
+				const guint8 *data,
+				guint datasz,
+				GError **error)
+{
+	fu_dump_raw(G_LOG_DOMAIN, "SetFeature", data, datasz);
+	return fu_udev_device_ioctl(FU_UDEV_DEVICE(self),
+				    HIDIOCSFEATURE(datasz),
+				    (guint8 *)data,
+				    NULL,
+				    FU_WEIDA_RAW_DEVICE_IOCTL_TIMEOUT,
+				    error);
+
+	return TRUE;
+}
+
+static gboolean
+fu_weida_raw_device_get_feature(FuWeidaRawDevice *self, guint8 *data, guint datasz, GError **error)
+{
+	if (!fu_udev_device_ioctl(FU_UDEV_DEVICE(self),
+				  HIDIOCGFEATURE(datasz),
+				  data,
+				  NULL,
+				  FU_WEIDA_RAW_DEVICE_IOCTL_TIMEOUT,
+				  error))
+		return FALSE;
+	fu_dump_raw(G_LOG_DOMAIN, "GetFeature", data, datasz);
+
+	return TRUE;
+}
+
+static void
+fu_weida_raw_device_to_string(FuDevice *device, guint idt, GString *str)
+{
+	return;
```

Comment:
you need to dump:

	gint32 dev_type;
	gint32 firmware_id;
	gint32 hardware_id;
	guint8 firmware_rev_ext;

here

Reply:
soloved